src/AspNetCore/Src/applicationmanager.cxx

```diff
@@ -225,24 +248,21 @@ APPLICATION_MANAGER::RecycleApplication(
     return hr;
 }
 
+//
+// Calls shutdown on all APPLICATION_INFO objects in the manager.
+// 
 VOID
-APPLICATION_MANAGER::ShutDown()
+APPLICATION_MANAGER::ShutDown(
+    VOID
+)
 {
     m_fInShutdown = TRUE;
     if (m_pApplicationInfoHash != NULL)
     {
         AcquireSRWLockExclusive(&m_srwLock);
 
-        // clean up the hash table so that the application will be informed on shutdown
-        m_pApplicationInfoHash->Clear();
+        m_pApplicationInfoHash->Apply(APPLICATION_INFO::ShutDownWrapper, NULL);
 
         ReleaseSRWLockExclusive(&m_srwLock);
     }
-
-    // stop filewatcher monitoring thread
```

Comment:
where do you clean up mpFileWatcher?

Reply:
In the destructor, but I think you are right.